classes/village/VillageManager.php

```diff
@@ -518,49 +532,66 @@ public static function getVillageSeats(System $system, int $village_id): array {
-    {
-        $resource_history = [];
+    public static function getResourceHistory(System $system, int $village_id, int $days): array {
+        $resource_history = [
+            WarManager::RESOURCE_MATERIALS => ['produced' => 0, 'collected' => 0, 'spent' => 0, 'claimed' => 0, 'lost' => 0],
+            WarManager::RESOURCE_FOOD => ['produced' => 0, 'collected' => 0, 'spent' => 0, 'claimed' => 0, 'lost' => 0],
+            WarManager::RESOURCE_WEALTH => ['produced' => 0, 'collected' => 0, 'spent' => 0, 'claimed' => 0, 'lost' => 0],
+        ];
         $time = time() - ($days * 86400);
-        foreach (array_keys(WarManager::RESOURCE_NAMES) as $resource_id) {
-            // get produced
-            $result = $system->db->query("SELECT SUM(`quantity`) as 'produced' FROM `resource_logs` WHERE `resource_id` = {$resource_id} AND `village_id` = {$village_id} AND `type` = " . self::RESOURCE_LOG_PRODUCTION . " AND `time` > {$time}");
-            $result = $system->db->fetch($result);
-            $resource_history[$resource_id]['produced'] = (int)$result['produced'];
-            // get collected
-            $result = $system->db->query("SELECT SUM(`quantity`) as 'collected' FROM `resource_logs` WHERE `resource_id` = {$resource_id} AND `village_id` = {$village_id} AND `type` = " . self::RESOURCE_LOG_COLLECTION . " AND `time` > {$time}");
-            $result = $system->db->fetch($result);
-            $resource_history[$resource_id]['collected'] = (int)$result['collected'];
-            // get claimed
-            $result = $system->db->query("SELECT COUNT(*) as 'claimed_loot_count' FROM `loot` WHERE `resource_id` = {$resource_id} AND `claimed_village_id` = {$village_id} AND `claimed_time` > {$time}");
-            $result = $system->db->fetch($result);
-            $resource_history[$resource_id]['claimed'] = (int)$result['claimed_loot_count'];
-            // get lost (lol)
-            $result = $system->db->query("SELECT COUNT(*) as 'lost_loot_count' FROM `loot` WHERE `resource_id` = {$resource_id} AND `target_village_id` = {$village_id} AND (`claimed_village_id` != {$village_id} OR `claimed_village_id` IS NULL) AND (`claimed_time` > {$time} OR `claimed_time` IS NULL)");
-            $result = $system->db->fetch($result);
-            $resource_history[$resource_id]['lost'] = (int) $result['lost_loot_count'];
-            // get upkeep - WIP
-            $result = $system->db->query("SELECT SUM(`quantity`) as 'spent' FROM `resource_logs` WHERE `resource_id` = {$resource_id} AND `village_id` = {$village_id} AND `type` = " . self::RESOURCE_LOG_EXPENSE . " AND `time` > {$time}");
-            $result = $system->db->fetch($result);
-            $resource_history[$resource_id]['spent'] = (int) $result['spent'];
+        $query = "
+            SELECT
+                `resource_id`,
+                SUM(CASE WHEN `type` = " . self::RESOURCE_LOG_PRODUCTION . " THEN `quantity` ELSE 0 END) AS 'produced',
```

Comment:
Just checking, do we have an index on `type` in the resource_logs table? How often will this function be called?

Reply:
We do not yet but that's part of optimizing this query. Code-wise changing this to hitting the logs table once rather than multiple times per resource, and then adding an index on type.